kyaml/fn/framework/selector.go

```diff
@@ -0,0 +1,207 @@
+// Copyright 2021 The Kubernetes Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+package framework
+
+import (
+	"sigs.k8s.io/kustomize/kyaml/errors"
+	"sigs.k8s.io/kustomize/kyaml/yaml"
+)
+
+// Selector matches resources.  A resource matches if and only if ALL of the Selector fields
+// match the resource.  An empty Selector matches all resources.
+type Selector struct {
+	// Names is a list of metadata.names to match.  If empty match all names.
+	// e.g. Names: ["foo", "bar"] matches if `metadata.name` is either "foo" or "bar".
+	Names []string `json:"names" yaml:"names"`
+
+	// Namespaces is a list of metadata.namespaces to match.  If empty match all namespaces.
+	// e.g. Namespaces: ["foo", "bar"] matches if `metadata.namespace` is either "foo" or "bar".
+	Namespaces []string `json:"namespaces" yaml:"namespaces"`
+
+	// Kinds is a list of kinds to match.  If empty match all kinds.
+	// e.g. Kinds: ["foo", "bar"] matches if `kind` is either "foo" or "bar".
+	Kinds []string `json:"kinds" yaml:"kinds"`
+
+	// APIVersions is a list of apiVersions to match.  If empty apply match all apiVersions.
+	// e.g. APIVersions: ["foo/v1", "bar/v1"] matches if `apiVersion` is either "foo/v1" or "bar/v1".
+	APIVersions []string `json:"apiVersions" yaml:"apiVersions"`
+
+	// Labels is a collection of labels to match.  All labels must match exactly.
+	// e.g. Labels: {"foo": "bar", "baz": "buz"] matches if BOTH "foo" and "baz" labels match.
+	Labels map[string]string `json:"labels" yaml:"labels"`
+
+	// Annotations is a collection of annotations to match.  All annotations must match exactly.
+	// e.g. Annotations: {"foo": "bar", "baz": "buz"] matches if BOTH "foo" and "baz" annotations match.
+	Annotations map[string]string `json:"annotations" yaml:"annotations"`
+
+	// ResourceMatcher is an arbitrary function used to match resources.
+	// Selector matches if the function returns true.
+	ResourceMatcher func(*yaml.RNode) bool
+
+	// TemplateData if present will cause the selector values to be parsed as templates
+	// and rendered using TemplateData before they are used.
+	TemplateData interface{}
+
+	// FailOnEmptyMatch makes the selector return an error when no items are selected.
+	FailOnEmptyMatch bool
+}
+
+// Filter implements kio.Filter, returning only those items from the list that the selector
+// matches.
+func (s *Selector) Filter(items []*yaml.RNode) ([]*yaml.RNode, error) {
+	andSel := AndSelector{TemplateData: s.TemplateData, FailOnEmptyMatch: s.FailOnEmptyMatch}
+	if s.Names != nil {
```

Comment:
This also works without the if statements, `s.Names...` works even when `s.Names` is nil.

Reply:
Actually, this could be:

```go
andSel := AndSelector {
  TemplateData: ...
  Matchers: []ResourceMatcher{
    NameMatcher(s.Names...),
    NamespaceMatcher(s.Namespaces...),
   ...
}
```

Reply:
True, but it feels weird/inefficient (even if not meaningfully less performant in practice) to me to put the resources through all those matchers unconditionally.